### ISSUE TITLE
Cannot open SSL keys

### ISSUE THREAD

**Issue by salvor-hardin** [Sequel-Ace/Sequel-Ace#52]
macOS 10.15.5
Sequel Ace 2.0.1

Cannot open SSH key files from anywhere

<img width="406" alt="0-06-22 at 11 41 09" src="https://user-images.githubusercontent.com/40695717/85278793-50ef6780-b47d-11ea-8b9b-2c53228d6fc9.png">

**Comment by gboudreau** [comment]
What is shown in the `SSH Key` field?
Please provide a screenshot like this:
![image](https://user-images.githubusercontent.com/370329/85279784-0528c800-b455-11ea-91a9-d49a23b37cdb.png)

**Comment by salvor-hardin** [comment]
Hi:

The file I am trying to read with the keys, same as my export from Sequel Pro.

I have this that works with Sequel Pro:

<img width="352" alt="2020-06-22 at 11 55 05" src="https://user-images.githubusercontent.com/40695717/85280287-bf352980-b47f-11ea-9689-80891b3e1c02.png">


When I try to connect, the key file gets removed as below and I get this:

<img width="488" alt="2020-06-22 at 11 55 17" src="https://user-images.githubusercontent.com/40695717/85280293-c0feed00-b47f-11ea-90df-c0ae36a5aaa1.png">



However the file is there, readable. Moved to different locations, same error. Given the tool full disk access, same error…

Thanks.

**Comment by gboudreau** [comment]
Ah, the SSL key/certs, not the SSH Key.
Another sandboxing issue.

**Comment by jamesstout** [comment]
I'm looking at this now

**Comment by chrismaaz** [comment]
Same problem here, SSH key can not be found when migrating from Sequel Pro, selecting a new one messes the path up.

**Comment by jamesstout** [comment]
Can someone please try out https://github.com/jamesstout/Sequel-Ace/tree/more-permissions

**Comment by mcdado** [comment]
In my case, after selecting what is ~/.ssh/my-key.pub, the link in the key field has the container path in it. Is it a problem with the actual private key not being in it?

**Comment by chrismaaz** [comment]
I get some kind of virtual path (I can't even see the full path), I guess a caveat of the app being in "sandbox mode" still?

**Comment by mcdado** [comment]
```
debug1: Server accepts key: /Users/David/Library/Containers/com.sequel-ace.sequel-ace/Data/.keys/123456789123456789123-my-key.pub RSA SHA256:..../..+..+.... explicit
Load key "/Users/David/Library/Containers/com.sequel-ace.sequel-ace/Data/.keys/123456789123456789123-my-key.pub": invalid format
debug1: No more authentication methods to try.
user@vps123456.ovh.net: Permission denied (publickey).
```

From the Error Detail screen. I anonymized the values here.

**Comment by Jason-Morcos** [comment]
@mcdado Are you certain the .pub key is correct? I know when I SSH into repos I use a .pem, not .pub key. Are you able to test with a .pem key and see if that works?

**Comment by Jason-Morcos** [comment]
> In my case, after selecting what is ~/.ssh/my-key.pub, the link in the key field has the container path in it. Is it a problem with the actual private key not being in it?

@mcdado @chrismaaz Yes, that's part of sandbox world. When you select a key, it's actually copied into the app's support directory and this copy is the one used for authentication. The path will indeed have all that wildness in it. Perhaps in the future we can make it a little prettier (or use bookmarks instead of copying @jamesstout?)!

**Comment by mcdado** [comment]
> @mcdado Are you certain the .pub key is correct? I know when I SSH into repos I use a .pem, not .pub key. Are you able to test with a .pem key and see if that works?

![Screenshot_2020-06-22_at_22_59_59](https://user-images.githubusercontent.com/898057/85334865-3f897800-b4dc-11ea-9b14-6fa5fa2e37b1.png)

I generated this key in this way: https://help.github.com/en/github/authenticating-to-github/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent

**Comment by gboudreau** [comment]
@mcdado You need to select the private key, not the public key. The public key is on the server, the private key is the one the client needs.

**Comment by gboudreau** [comment]
(We really should not let users select .pub files...)

**Comment by mcdado** [comment]
> @mcdado You need to select the private key, not the public key. The public key is on the server, the private key is the one the client needs.

Sure... in Sequel Pro (Nightly too) i always selected the pub, I guess I'm used to the -i flag in ssh.

**Comment by mcdado** [comment]
Even with the private key it doesn't work.

**Comment by gboudreau** [comment]
`ssh -i` is actually for the private key too, not the public key:
```
     -i identity_file
             Selects a file from which the identity (private key) for public key authentication is read.
```
I think `ssh` (the executable) is smart enough that if you specify a `.pub` file, it will actually try to use the associated private key (same filename, minus the `.pub`). Sequel Ace can't do that, because it only has access to the files you select in the picker, nothing else (because of sandboxing).

Is your SSH private key encrypted? Are you using Sequel Ace 2.0.2 or later?
If your SSH key is encrypted, Sequel Ace will need to ask you for the password to decrypt it, and 2.0.1 and earlier versions have a bug that prevented that.

**Comment by mcdado** [comment]
I'm using 2.0.1 (just downloaded from MAS), my SSH has a passphrase but ssh-key agent unlocks it.

**Comment by stychos** [comment]
@mcdado try newer version from [Releases page](https://github.com/Sequel-Ace/Sequel-Ace/releases) or wait until 2.0.2 release in MAS.

**Comment by Jason-Morcos** [comment]
2.0.2 is pending App Store review, but they're never very quick to review new releases. Could be a couple days until out.

**Comment by mcdado** [comment]
Yes, 2.0.2 works! It shows a dropdown sheet for password input.

Out of curiosity: because of sandboxing it cannot use ssh-agent to unlock ssh keys?

**Comment by Jason-Morcos** [comment]
@mcdado Probably can't? Not sure. Sandbox is _very_ restricting, but I guess that's kinda the point lol

**Comment by jamesstout** [comment]
> @mcdado @chrismaaz Yes, that's part of sandbox world. When you select a key, it's actually copied into the app's support directory and this copy is the one used for authentication. The path will indeed have all that wildness in it. Perhaps in the future we can make it a little prettier (or use bookmarks instead of copying @jamesstout?)!

You can add key-values to the bookmark (I'm not doing that at the minute), we could probably add the original path/filename to the bookmark and then populate the field with that, or just try the bookmark without copying.

**Comment by jamesstout** [comment]
What do you prefer:

1. Populate original path but use the copied key - quicker fix, but might confuse more advanced users.
2. Try bookmarking without opting to `.keys`? - slightly longer fix I think, but cleaner.

**Comment by Jason-Morcos** [comment]
> What do you prefer:
> 
> 
> 
> 1. Populate original path but use the copied key - quicker fix, but might confuse more advanced users.
> 
> 2. Try bookmarking without opting to `.keys`? - slightly longer fix I think, but cleaner.

I think 2 is probably the better move long-term?

**Comment by jamesstout** [comment]
I've implemented 2 in https://github.com/jamesstout/Sequel-Ace/tree/more-permissions.

**Comment by Jason-Morcos** [comment]
@jamesstout your branch looks great to me! At least worked great locally. PR?

**Comment by jamesstout** [comment]
Yep, will just sort my branch out and submit.

**Comment by jamesstout** [comment]
PR https://github.com/Sequel-Ace/Sequel-Ace/pull/105

**Comment by Jason-Morcos** [comment]
Please try the latest beta - this issue should be fixed and will be patched in the 2.1.0 release! https://github.com/Sequel-Ace/Sequel-Ace/releases/tag/2.1.0-rc1

**Comment by salvor-hardin** [comment]
Hi:

This issue is now fixed with version 2.1.0

Thanks!
